### ISSUE TITLE
[Proposal] Add `On Return` section at the end of functions and subs

### ISSUE THREAD

**Issue by VBAndCs** [dotnet/vblang#569]
In the a function named Validate, I needed to use a flag to prevent endless LostFocus event firing, so, I ended up with this code 
```VB.NET
Dim IsValidating As Boolean
Function Validate() As Boolean
        IsValidating = True

        If cmbTransferType.HasError Then
            cmbTransferType.Focus()
            IsValidating = False
            Return False
        End If

        If Not ValidateNumber() Then
            txtNumber.Focus()
            IsValidating = False
            Return False
        End If

        If txtAmount.HasErrors Then
            txtAmount.Focus()
            IsValidating = False
            Return False
        End If

        If txtCharge.HasErrors Then
            txtCharge.Focus()
            IsValidating = False
            Return False
        End If

        If Not ValidateValidDate() Then
            dtpValidDate.Focus()
            IsValidating = False
            Return False
        End If

        If cmbState.HasError Then
            cmbState.Focus()
            IsValidating = False
            Return False
        End If

        If Not ValidateCashedDate() Then
            dtpCashedDate.Focus()
            IsValidating = False
            Return False
        End If

        IsValidating = False
        Return True


End Function
```

As you can see, I pasted `IsValidating = False` 8 times (before every return statement).
I wish to have something to enhance this style, without having to use goto statements, say a `On Return` section at the end of the function, where we can execute what ever we want before the function returns:

```VB.NET
Dim IsValidating As Boolean
    Function Validate() As Boolean
        IsValidating = True

        If cmbTransferType.HasError Then
            cmbTransferType.Focus()
            Return False
        End If

        ' …………………… The rest of the code

        Return True

   On Return:
      IsValidating = False        
End Function
```

The above code can be lowered to this classical syntax using goto and the name of the function as a store for the return value, which is less readable and a bit longer:
```VB.NET
Dim IsValidating As Boolean
    Function Validate() As Boolean
        IsValidating = True

        If cmbTransferType.HasError Then
            cmbTransferType.Focus()
            Validate = False
            Goto OnReturn:
        End If

        ' …………………… The rest of the code

        Validate = True

   OnReturn:
      IsValidating = False        
End Function
```

I think using the suggested `On Return` section will enhance functions code, making it easy to modify it. For example, if I decide to get rid of the flag or to add more statements, It will be a minimal change, in a one section at the end of the function.

**Comment by zspitz** [comment]
If you want to avoid boilerplate, you could restructure your code a bit:
```vb
Dim validators = New List(Of (Func(Of Boolean), Control)) From {
    (Function() Not cmbTransfers.HasError, cmbTransferType),
    (Function() ValidateNumber, txtNumber),
    (Function() Not txtAmount.HasErrors, txtAmount),
    (Function() Not txtCharge.HasErrors, txtCharge),
    (Function() ValidateValidDate, dtpValidDate),
    (Function() Not cmbState.HasErrors, cmbState),
    (Function() ValidateCashedDate, dtpCashedDate)
}

Dim IsValidating As Boolean

Function Validate() As Boolean
    IsValidating = True
    Dim isValid As Boolean = True
    For Each pair In validators
        If pair.Item1.Invoke Then Continue For
        pair.Item2.Focus
        isValid = False
        Exit For
    Next
    IsValidating = False
    Return isValid
End Function
```

**Comment by pricerc** [comment]
Or how about just a good old try... finally (without the catch).
``` VB.Net
Dim IsValidating As Boolean
Function Validate() As Boolean
    Try
        IsValidating = True

        If cmbTransferType.HasError Then
            cmbTransferType.Focus()
            Return False
        End If

        If Not ValidateNumber() Then
            txtNumber.Focus()
            Return False
        End If

        If txtAmount.HasErrors Then
            txtAmount.Focus()
            Return False
        End If

        If txtCharge.HasErrors Then
            txtCharge.Focus()
            Return False
        End If

        If Not ValidateValidDate() Then
            dtpValidDate.Focus()
            Return False
        End If

        If cmbState.HasError Then
            cmbState.Focus()
            Return False
        End If

        If Not ValidateCashedDate() Then
            dtpCashedDate.Focus()
            Return False
        End If

        Return True
    Finally
        IsValidating = False
    End Try

End Function
```

**Comment by VBAndCs** [comment]
The `try` solution is nice (yet a little longer and involves extra indentation). I never thought to use finally this way. I always think of it as a place to dispose resources or alike.
Still, I find this proposal useful, and easy to implement, with no penalties.

**Comment by Happypig375** [comment]
>Still, I find this proposal useful, and easy to implement, with no penalties.

You can see how hard it actually is by implementing yourself.

**Comment by VBAndCs** [comment]
@pricerc
I used your solution in another use case, but it has an issue: you can't use return from the finally block, so I needed to use a flag in this case (using the function name is not suitable for my case). `On Return` will not have this issue.

**Comment by VBAndCs** [comment]
@pricerc
I used your solution in another use case, but it has an issue: you can't use return from the finally block, so I needed to use a flag in this case (using the function name is not suitable for my case). `On Return` will not have this issue.

**Comment by hartmair** [comment]
LINQ version of @zspitz approach with try-finally block.
```vb
ReadOnly validators As (ValidateFunc As Func(Of Boolean), Control As Control)() = {
(Function() Not cmbTransferType.HasError, cmbTransferType),
(AddressOf ValidateNumber, txtNumber),
(Function() Not txtAmount.HasErrors, txtAmount),
(Function() Not txtCharge.HasErrors, txtCharge),
(AddressOf ValidateValidDate, dtpValidDate),
(Function() Not cmbState.HasErrors, cmbState),
(AddressOf ValidateCashedDate, dtpCashedDate)
}
Dim IsValidating As Boolean
Function Validate() As Boolean
    IsValidating = True
    Try
        Dim invalid = validators.FirstOrDefault(Function(a) Not a.ValidateFunc()).Control
        invalid?.Focus()
        Return invalid Is Nothing
    Finally
        IsValidating = False
    End Try
End Function
```

**Comment by VBAndCs** [comment]
I used another approach for validation. I found it easier and time saving to inherit the wpf controls to add a built in validation, so, I can reuse them in many windows on the project with minimal code. I am already using the IoC lambda approach to deal with the properties of the data source (instead of using reflection). I make my controls implement an IValidate interface, so I can use an extension method on the window to call the Validate method for all of them (and other methods). I provide extra validation through `IValidate.ValidateFunc`.
I hope WinUI provide such a built in capability.  
But I am still facing other situations where I need the `On Return` section to minimize code or simplify the nested Ifs. I still think it is a good thing to have.
